File: Sqrilizz-Reports/src/main/java/dev/sqrilizz/SQRILIZZREPORTS/MySQLOptionalManager.java
package dev.sqrilizz.SQRILIZZREPORTS;

import java.sql.Connection;
import java.sql.Driver;
import java.sql.DriverManager;
import java.sql.SQLException;

/**
 * Optional MySQL manager that loads MySQL driver dynamically
 * This allows the plugin to work without MySQL JAR in the main plugin file
 * Users can put mysql-connector-j.jar in plugins/Sqrilizz-Reports/lib/ folder
 */
public class MySQLOptionalManager {
    private static boolean mysqlAvailable = false;
    private static String mysqlDriverPath = null;
    
    static {
        checkMySQLAvailability();
    }
    
    private static void checkMySQLAvailability() {
        try {
            // Try to load MySQL driver from classpath first
            Class.forName("com.mysql.cj.jdbc.Driver");
            mysqlAvailable = true;
            Main.getInstance().getLogger().info("MySQL driver found in classpath");
        } catch (ClassNotFoundException e) {
            // Try to find MySQL driver in plugins folder
            java.io.File pluginFolder = Main.getInstance().getDataFolder();
            java.io.File libFolder = new java.io.File(pluginFolder, "lib");
            
            if (libFolder.exists()) {
                java.io.File[] jarFiles = libFolder.listFiles((dir, name) -> 
                    name.toLowerCase().contains("mysql") && name.endsWith(".jar"));
                
                if (jarFiles != null && jarFiles.length > 0) {
                    try {
                        // Load MySQL driver from external JAR
                        java.net.URL jarUrl = jarFiles[0].toURI().toURL();
                        java.net.URLClassLoader classLoader = new java.net.URLClassLoader(new java.net.URL[]{jarUrl});
                        Class<?> driverClass = classLoader.loadClass("com.mysql.cj.jdbc.Driver");
                        Driver driver = (Driver) driverClass.getDeclaredConstructor().newInstance();
                        DriverManager.registerDriver(driver);
                        
                        mysqlAvailable = true;
                        mysqlDriverPath = jarFiles[0].getAbsolutePath();
                        Main.getInstance().getLogger().info("MySQL driver loaded from: " + mysqlDriverPath);
                    } catch (Exception ex) {
                        Main.getInstance().getLogger().warning("Failed to load external MySQL driver: " + ex.getMessage());
                    }
                }
            }
            
            if (!mysqlAvailable) {
                Main.getInstance().getLogger().info("MySQL driver not found. Plugin will use SQLite only.");
                Main.getInstance().getLogger().info("To enable MySQL support, place mysql-connector-j.jar in plugins/Sqrilizz-Reports/lib/ folder");
            }
        }
    }
    
    public static boolean isMySQLAvailable() {
        return mysqlAvailable;
    }
    
    public static Connection createMySQLConnection(String url, String username, String password) throws SQLException {
        if (!mysqlAvailable) {
            throw new SQLException("MySQL driver is not available");
        }
        
        return DriverManager.getConnection(url, username, password);
    }
    
    public static String getMySQLDriverInfo() {
        if (!mysqlAvailable) {
            return "MySQL driver not available";
        }
        
        if (mysqlDriverPath != null) {
            return "External MySQL driver: " + mysqlDriverPath;
        } else {
            return "Built-in MySQL driver";
        }
    }
    
    /**
     * Creates lib folder and provides instructions for MySQL setup
     */
    public static void setupMySQLInstructions() {
        java.io.File pluginFolder = Main.getInstance().getDataFolder();
        java.io.File libFolder = new java.io.File(pluginFolder, "lib");
        
        if (!libFolder.exists()) {
            libFolder.mkdirs();
            
            // Create README file with instructions
            java.io.File readmeFile = new java.io.File(libFolder, "README.txt");
            try (java.io.FileWriter writer = new java.io.FileWriter(readmeFile)) {
                writer.write("MySQL Support Setup Instructions\n");
                writer.write("================================\n\n");
                writer.write("To enable MySQL support for Sqrilizz-Reports:\n\n");
                writer.write("1. Download mysql-connector-j.jar from:\n");
                writer.write("   https://dev.mysql.com/downloads/connector/j/\n\n");
                writer.write("2. Place the JAR file in this folder (plugins/Sqrilizz-Reports/lib/)\n\n");
                writer.write("3. Restart the server\n\n");
                writer.write("The plugin will automatically detect and load the MySQL driver.\n");
                writer.write("If MySQL driver is not found, the plugin will use SQLite instead.\n");
            } catch (Exception e) {
                Main.getInstance().getLogger().warning("Could not create MySQL setup instructions: " + e.getMessage());
            }
        }
    }
}

File: Sqrilizz-Reports/src/main/kotlin/dev/sqrilizz/SQRILIZZREPORTS/MySQLOptionalManager.kt
package dev.sqrilizz.SQRILIZZREPORTS

import java.io.File
import java.io.FileWriter
import java.net.URL
import java.net.URLClassLoader
import java.sql.Connection
import java.sql.Driver
import java.sql.DriverManager
import java.sql.SQLException

object MySQLOptionalManager {

    private var mysqlAvailable = false
    private var mysqlDriverPath: String? = null

    init {
        checkMySQLAvailability()
    }

    private fun checkMySQLAvailability() {
        try {
            Class.forName("com.mysql.cj.jdbc.Driver")
            mysqlAvailable = true
            Main.getInstance().logger.info("MySQL driver found in classpath")
        } catch (_: ClassNotFoundException) {
            val pluginFolder = Main.getInstance().dataFolder
            val libFolder = File(pluginFolder, "lib")

            if (libFolder.exists()) {
                val jarFiles = libFolder.listFiles { _, name ->
                    name.lowercase().contains("mysql") && name.endsWith(".jar")
                }

                if (!jarFiles.isNullOrEmpty()) {
                    try {
                        val jarUrl = jarFiles[0].toURI().toURL()
                        val classLoader = URLClassLoader(arrayOf(jarUrl))
                        val driverClass = classLoader.loadClass("com.mysql.cj.jdbc.Driver")
                        val driver = driverClass.getDeclaredConstructor().newInstance() as Driver
                        DriverManager.registerDriver(driver)

                        mysqlAvailable = true
                        mysqlDriverPath = jarFiles[0].absolutePath
                        Main.getInstance().logger.info("MySQL driver loaded from: $mysqlDriverPath")
                    } catch (ex: Exception) {
                        Main.getInstance().logger.warning("Failed to load external MySQL driver: ${ex.message}")
                    }
                }
            }

            if (!mysqlAvailable) {
                Main.getInstance().logger.info("MySQL driver not found. Plugin will use SQLite only.")
                Main.getInstance().logger.info("To enable MySQL support, place mysql-connector-j.jar in plugins/Sqrilizz-Reports/lib/ folder")
            }
        }
    }

    @JvmStatic
    fun isMySQLAvailable(): Boolean = mysqlAvailable

    @JvmStatic
    @Throws(SQLException::class)
    fun createMySQLConnection(url: String, username: String, password: String): Connection {
        if (!mysqlAvailable) {
            throw SQLException("MySQL driver is not available")
        }
        return DriverManager.getConnection(url, username, password)
    }

    @JvmStatic
    fun getMySQLDriverInfo(): String = when {
        !mysqlAvailable -> "MySQL driver not available"
        mysqlDriverPath != null -> "External MySQL driver: $mysqlDriverPath"
        else -> "Built-in MySQL driver"
    }

    @JvmStatic
    fun setupMySQLInstructions() {
        val pluginFolder = Main.getInstance().dataFolder
        val libFolder = File(pluginFolder, "lib")

        if (!libFolder.exists()) {
            libFolder.mkdirs()

            val readmeFile = File(libFolder, "README.txt")
            try {
                FileWriter(readmeFile).use { writer ->
                    writer.write("""
                        |MySQL Support Setup Instructions
                        |================================
                        |
                        |To enable MySQL support for Sqrilizz-Reports:
                        |
                        |1. Download mysql-connector-j.jar from:
                        |   https://dev.mysql.com/downloads/connector/j/
                        |
                        |2. Place the JAR file in this folder (plugins/Sqrilizz-Reports/lib/)
                        |
                        |3. Restart the server
                        |
                        |The plugin will automatically detect and load the MySQL driver.
                        |If MySQL driver is not found, the plugin will use SQLite instead.
                    """.trimMargin())
                }
            } catch (_: Exception) {
                // ignore
            }
        }
    }
}
